### PR TITLE
`Development`: Handle course image persistence directly

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

In #5733 I replaced the two-parted upload with a single multipart upload but forgot to remove the delayed persistence of the file path that was now obsolete.

### Description

In this PR, I opened the API of the file service to allow services to save a file directly to a path instead of saving it to a temporary path as before. I removed the delayed persistence method from the Course object and let the course resources handle the file saving directly.
Creating a course requires an additional database call as the current implementation requires the ID as part of the course Icon public path. There are ways to prevent this, but it requires a completely different implementation or to keep replacing the ID in the path on load.

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Instructor

1. Create Course A with an Admin with a custom image
2. Create Course B with an Admin without a custom image
3. Update a course with an admin adapting the image (expect image updated)
4. Update a course with an admin adapting the user groups (expect groups updated)
5. Update a course with an instructor adapting the image (expect image updated)
6. Delete the courses

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| EntityFileService.java | 97.0% |                            |
| FileService.java | 86.1% |               |
| LectureUnitProcessingService.java | 91.45% |               |
| QuizExerciseService.java | 97.6% |               |
| CourseResource.java | 96.0% |               |
| AdminCourseResource.java | 98.5% |               |

Only changes that are expected to require no test adaptions. All changes in FileService (<90%) have enough coverage.

## Summary by CodeRabbit

- **Refactor**
	- Simplified file management in course-related operations.
	- Enhanced file path generation logic for better consistency.
	- Improved file saving process with clearer method responsibilities.

- **Bug Fixes**
	- Ensured file names are safely generated and handled.

- **New Features**
	- Introduced null safety checks for file operations.

- **Documentation**
	- Updated method documentation to reflect changes in file handling logic.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 1]
```diff
@@ -178,20 +174,52 @@
...
+     */
+    @NotNull
+    public Path saveFileFoo(MultipartFile file, Path basePath) {
+        String sanitizedFilename = checkAndSanitizeFilename(file.getOriginalFilename());
+        validateExtension(sanitizedFilename, false);
+        String generatedFilename = generateFilename(generateTargetFilenameBase(basePath), FilenameUtils.getExtension(sanitizedFilename));
+        Path savePath = basePath.resolve(generatedFilename);
+        copyFile(file, savePath);
+        return savePath;
+    }
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java]: Fixed in 6e09f0ce5d39356f6120e10bd9d2a5afc2ed22ea

**Review by konrad2002 — APPROVED** [review]
manually tested on ts1, works as expected

**Review by MaximilianJG — APPROVED** [review]
Tested on Ts2. Works as expected

**Review by MichaelOwenDyer — APPROVED** [review]
Code looks clean and I don't see any issues 👍🏽

**Pull request merged**